chat-android/src/main/java/com/ably/chat/Message.kt

```diff
@@ -68,10 +71,108 @@ data class Message(
      */
-    val headers: MessageHeaders,
+    val headers: MessageHeaders?,
 
     /**
      * The latest action of the message. This can be used to determine if the message was created, updated, or deleted.
+     * Spec: CHA-M10
      */
     val action: MessageAction,
+
+    /**
+     * A unique identifier for the latest version of this message.
+     * Spec: CHA-M10a
+     */
+    val version: String,
+
+    /**
+     * The timestamp at which this version was updated, deleted, or created.
+     */
+    val timestamp: Long,
+
+    /**
+     * The details of the operation that modified the message. This is only set for update and delete actions. It contains
+     * information about the operation: the clientId of the user who performed the operation, a description, and metadata.
+     */
+    val operation: Message.Operation? = null,
 )
+
+fun com.ably.chat.Message.copy(text: String, metadata: MessageMetadata? = null, headers: MessageHeaders? = null): MessageCopy {
+    return MessageCopy(
+        message = this,
+        text = text,
+        metadata = metadata,
+        headers = headers,
+    )
+}
+
+data class MessageCopy(
+    val message: com.ably.chat.Message,
+    val text: String,
+    val metadata: MessageMetadata? = null,
+    val headers: MessageHeaders? = null,
+) {
+    val serial: String = message.serial
+    val clientId: String = message.clientId
+    val roomId: String = message.roomId
+    val createdAt: Long = message.createdAt
+    val action: MessageAction = message.action
+    val version: String = message.version
+    val timestamp: Long = message.timestamp
+    val operation: Message.Operation? = message.operation
+}
+
+fun buildMessageOperation(jsonObject: JsonObject?): Message.Operation? {
+    if (jsonObject == null) {
+        return null
+    }
+    val operation = Message.Operation()
+    if (jsonObject.has(MessageOperationProperty.CLIENT_ID)) {
+        operation.clientId = jsonObject.get(MessageOperationProperty.CLIENT_ID).asString
+    }
+    if (jsonObject.has(MessageOperationProperty.DESCRIPTION)) {
+        operation.description = jsonObject.get(MessageOperationProperty.DESCRIPTION).asString
+    }
+    if (jsonObject.has(MessageOperationProperty.METADATA)) {
+        val metadataObject = jsonObject.getAsJsonObject(MessageOperationProperty.METADATA)
+        operation.metadata = mutableMapOf()
+        for ((key, value) in metadataObject.entrySet()) {
+            operation.metadata[key] = value.asString
+        }
+    }
+    return operation
+}
+
+fun buildMessageOperation(clientId: String, description: String?, metadata: Map<String, String>?): Message.Operation {
+    val operation = Message.Operation()
+    operation.clientId = clientId
+    operation.description = description
+    operation.metadata = metadata
+    return operation
+}
+
+/**
+ * MessageProperty object representing the properties of a message.
+ */
+object MessageProperty {
```

Comment:
It should be internal, I believe, and we should stick to PascalCase. Screaming SNAKE_CASE is appropriate for top-level constants and companion object constants. However, in this case, we have a set of properties with the same origin (similar to an enum), and I think it’s more common to use PascalCase. See https://github.com/ktorio/ktor/blob/348033b64c144090dacdecbf441f6c61826dca76/ktor-http/common/src/io/ktor/http/auth/AuthScheme.kt#L12 for example